### PR TITLE
fix container vm invalid state issue

### LINKED ISSUE #2698: faild to start docker container after vm enter into invalid state
Upgrade tests hit this while trying to power a VM back on.

https://ci.vmware.run/vmware/vic/5693

As this is an issue different with #1823, open a new issue here.
@hickeng @chengwang86

### PR DESCRIPTION
**Pull request by emlin**
Fixes #2698 

If any vm operation returned invalid state error, will try to unregister and register vm, and then retry task.
As unregister vm will remove container from portlayer container cache, add event listening for vm registered event.

### CODE REVIEW

**Comment by emlin** [comment]
@jzt

File: lib/portlayer/exec/exec.go [hunk 1]
```diff
@@ -152,8 +158,36 @@ func eventCallback(ie events.Event) {
...
+		attrib := []string{"config", "runtime.powerState", "summary", "resourcePool"}
 
+		// populate the vm properties
+		ctx := context.Background()
+		if err := eventSession.RetrieveOne(ctx, *moref, attrib, &vm); err != nil {
+			log.Errorf("Failed to query registered vm object %s: %s", ie.Reference(), err)
+			return
+		}
+		if *vm.ResourcePool != Config.ResourcePool.Reference() {
+			log.Debugf("vm %q is not containers of this VCH, ignore it", vm.Config.Name)
```
Comment (jzt) [inline lib/portlayer/exec/exec.go]: Minor: I would reword this error to say "`container vm %q does not belong to this VCH, ignoring`" or something.

**Comment by jzt** [comment]
LGTM

File: lib/portlayer/exec/exec.go (hunk 1 above)
Comment (emlin) [inline lib/portlayer/exec/exec.go]: nice, thanks

File: lib/portlayer/exec/exec.go [hunk 2]
```diff
@@ -152,8 +158,36 @@ func eventCallback(ie events.Event) {
...
+		}
+		registeredContainers := convertInfraContainers(ctx, eventSession, []mo.VirtualMachine{vm})
+		for i := range registeredContainers {
+			Containers.put(registeredContainers[i])
+			log.Debugf("Container %q is registered back", vm.Config.Name)
```
Comment (emlin) [inline lib/portlayer/exec/exec.go]: yes, sure

File: lib/portlayer/exec/exec.go [hunk 3]
```diff
@@ -152,8 +158,36 @@ func eventCallback(ie events.Event) {
...
 			}
 		}
+		return
 	}
+	// check container registered event if this container is not found in container cache
+	switch ie.String() {
+	case events.ContainerRegistered:
+		moref := new(types.ManagedObjectReference)
+		if ok := moref.FromString(ie.Reference()); !ok {
+			log.Errorf("Failed to get event VM mob reference")
```
Comment (vburenin) [inline lib/portlayer/exec/exec.go]: I would log reference here.

File: pkg/vsphere/vm/vm.go [hunk 4]
```diff
@@ -310,17 +310,91 @@ func (vm *VirtualMachine) DeleteExceptDisks(ctx context.Context) (*object.Task,
...
+
+	if err = vm.Properties(ctx, vm.Reference(), []string{"config.files"}, &mvm); err != nil {
+		log.Errorf("Unable to get vm config.files property: %s", err)
+		return "", err
+	}
+
+	return mvm.Config.Files.VmPathName, nil
+}
+
+func (vm *VirtualMachine) registerVM(ctx context.Context, path string, name string, vapp *types.ManagedObjectReference,
```
Comment (vburenin) [inline pkg/vsphere/vm/vm.go]: So many parameters for one function. Is it really necessary? You could also make definition much shorter combining types together.
```
registerVM(ctx context.Context, path, name string, vapp, pool, host *typesManagedObjectReference, folder *object.Folder)
```

btw, Why folder? Why not path?

File: pkg/vsphere/vm/vm.go [hunk 5]
```diff
@@ -310,17 +310,91 @@ func (vm *VirtualMachine) DeleteExceptDisks(ctx context.Context) (*object.Task,
...
+// FixInvalidState fix vm invalid state issue through unregister & register
+func (vm *VirtualMachine) FixInvalidState(ctx context.Context) error {
+	log.Debugf("Fix invalid state VM: %s", vm.Reference())
+	folders, err := vm.Session.Datacenter.Folders(ctx)
+	if err != nil {
+		log.Errorf("Unable to get vm folder: %s", err)
+		return err
+	}
+
+	log.Debugf("Get vm properties")
```
Comment (vburenin) [inline pkg/vsphere/vm/vm.go]: I would log which properties I am trying to retrieve. Taking into account you already have an array of the already, it should be trivial.

File: pkg/vsphere/vm/vm.go (hunk 4 above)
Comment (emlin) [inline pkg/vsphere/vm/vm.go]: yeah, it's necessary. with combined types, it will be better

File: pkg/vsphere/vm/vm.go (hunk 5 above)
Comment (emlin) [inline pkg/vsphere/vm/vm.go]: folder is actually vmfolder, will rename it

File: pkg/vsphere/vm/vm.go (hunk 4 above)
Comment (emlin) [inline pkg/vsphere/vm/vm.go]: folder is actually vmfolder, will rename it

File: pkg/vsphere/vm/vm.go (hunk 5 above)
Comment (emlin) [inline pkg/vsphere/vm/vm.go]: sure

File: lib/portlayer/exec/exec.go (hunk 3 above)
Comment (emlin) [inline lib/portlayer/exec/exec.go]: good point

File: pkg/vsphere/tasks/waiter.go [hunk 6]
```diff
@@ -89,6 +102,56 @@ func WaitForResult(ctx context.Context, f func(context.Context) (Task, error)) (
...
+	if !ok {
+		return false
+	}
+	switch f.Fault().(type) {
+	case *types.InvalidState:
+	default:
+		log.Debugf("Do not fix non invalid state error")
+		return false
+	}
+	if target == nil {
```
Comment (hmahmood) [inline pkg/vsphere/tasks/waiter.go]: You could check this first thing in this function.

File: pkg/vsphere/tasks/waiter.go [hunk 7]
```diff
@@ -89,6 +102,56 @@ func WaitForResult(ctx context.Context, f func(context.Context) (Task, error)) (
 	}
 }
 
+func fixTask(ctx context.Context, sess *session.Session, target object.Reference) error {
+	switch vmm := target.(type) {
+	case *object.VirtualMachine:
+		vm := vm.NewVirtualMachine(ctx, sess, target.Reference())
```
Comment (hmahmood) [inline pkg/vsphere/tasks/waiter.go]: You should get rid of the `sess` parameter to `NewVirtualMachine`. Its leaking badly into this API, and the `vm.VirtualMachine` class really uses it in only two places -- both of which can be replaced with a call to [`Client()`](https://github.com/vmware/vic/blob/master/vendor/github.com/vmware/govmomi/object/common.go#L62).

File: pkg/vsphere/tasks/waiter.go [hunk 8]
```diff
@@ -41,21 +46,21 @@ type Task interface {
...
 // WaitForResult wraps govmomi operations and wait the operation to complete.
 // Return the operation result
 // Sample usage:
-//    info, err := WaitForResult(ctx, func(ctx) (*TaskInfo, error) {
-//       return vm.Reconfigure(ctx, config)
+//    info, err := WaitForResult(ctx, sess *session.Session, *object.Reference, func(ctx) (*TaskInfo, error) {
+//       return vm, vm.Reconfigure(ctx, config)
 //    })
-func WaitForResult(ctx context.Context, f func(context.Context) (Task, error)) (*types.TaskInfo, error) {
+func WaitForResult(ctx context.Context, sess *session.Session, target object.Reference, f func(context.Context) (Task, error)) (*types.TaskInfo, error) {
```
Comment (hmahmood) [inline pkg/vsphere/tasks/waiter.go]: `TaskInfo.Entity` property holds the "target" of the task, so you can get rid of that parameter as well. See https://www.vmware.com/support/developer/vc-sdk/visdk41pubs/ApiReference/vim.TaskInfo.html and https://sourcegraph.com/github.com/vmware/vic@master/-/blob/vendor/github.com/vmware/govmomi/vim25/types/types.go#L39888

File: pkg/vsphere/tasks/waiter.go [hunk 9]
```diff
@@ -70,10 +75,18 @@ func WaitForResult(ctx context.Context, f func(context.Context) (Task, error)) (
 		}
 
 		log.Errorf("task failed: %s", err)
-		if !isTaskInProgress(err) {
+		if !needsRetry(target, err) {
```
Comment (hmahmood) [inline pkg/vsphere/tasks/waiter.go]: `needsRetry` calls `needsFix`. Maybe not have `needsRetry` call `needsFix` and instead call them separately here, caching the results for later reuse?

File: pkg/vsphere/vm/vm.go [hunk 10]
```diff
@@ -310,17 +310,91 @@ func (vm *VirtualMachine) DeleteExceptDisks(ctx context.Context) (*object.Task,
...
+	}
+
+	return mvm.Config.Files.VmPathName, nil
+}
+
+func (vm *VirtualMachine) registerVM(ctx context.Context, path, name string,
+	vapp, pool, host *types.ManagedObjectReference, vmfolder *object.Folder) (*object.Task, error) {
+	log.Debugf("Register VM %s", name)
+
+	c := vm.Client.Client
```
Comment (hmahmood) [inline pkg/vsphere/vm/vm.go]: `vm.Client()`?

File: pkg/vsphere/vm/vm.go [hunk 11]
```diff
@@ -310,17 +310,91 @@ func (vm *VirtualMachine) DeleteExceptDisks(ctx context.Context) (*object.Task,
...
 	if err != nil {
+		return nil, err
+	}
+
+	return object.NewTask(vm.Client.Client, res.Returnval), nil
+}
+
+// FixInvalidState fix vm invalid state issue through unregister & register
+func (vm *VirtualMachine) FixInvalidState(ctx context.Context) error {
+	log.Debugf("Fix invalid state VM: %s", vm.Reference())
```
Comment (hmahmood) [inline pkg/vsphere/vm/vm.go]: A sanity check to see if the VM is indeed in an invalid state before attempting to unregister would be nice.

File: pkg/vsphere/tasks/waiter.go [hunk 12]
```diff
@@ -70,10 +75,18 @@ func WaitForResult(ctx context.Context, f func(context.Context) (Task, error)) (
...
 			return info, err
 		}
 
+		if needsFix(target, err) {
+			if nerr := fixTask(ctx, sess, target); nerr != nil {
+				log.Errorf("Failed to fix task failure: %s", nerr)
```
Comment (hmahmood) [inline pkg/vsphere/tasks/waiter.go]: So no retry if we get a "task in progress" error?

File: pkg/vsphere/vm/vm.go [hunk 13]
```diff
@@ -310,17 +310,91 @@ func (vm *VirtualMachine) DeleteExceptDisks(ctx context.Context) (*object.Task,
...
+		return err
+	}
+	// re-register vm will change vm reference, so reset the object reference here
+	if info.Error != nil {
+		return errors.New(info.Error.LocalizedMessage)
+	}
+
+	newRef := info.Result.(types.ManagedObjectReference)
+
+	vm.Common = object.NewCommon(vm.Client.Client, newRef)
```
Comment (hmahmood) [inline pkg/vsphere/vm/vm.go]: Not `object.NewVirtualMachine`? There is an assumption here that `object.NewVirtualMachine` just creates `Common`, and if that changes in the future, we break here.

File: lib/portlayer/exec/exec.go [hunk 14]
```diff
@@ -152,8 +158,36 @@ func eventCallback(ie events.Event) {
 
 			}
 		}
+		return
 	}
+	// check container registered event if this container is not found in container cache
+	switch ie.String() {
```
Comment (hmahmood) [inline lib/portlayer/exec/exec.go]: Can you explain why this is required to fix the invalid state issue?

File: pkg/vsphere/tasks/waiter.go (hunk 12 above)
Comment (emlin) [inline pkg/vsphere/tasks/waiter.go]: this method is not retry, it's only fix the object. task will retry in existing way

File: pkg/vsphere/tasks/waiter.go (hunk 9 above)
Comment (emlin) [inline pkg/vsphere/tasks/waiter.go]: I feel it's more clear to have one single retry detection method. And the needsFix detection is easy to do.

File: pkg/vsphere/tasks/waiter.go (hunk 7 above)
Comment (emlin) [inline pkg/vsphere/tasks/waiter.go]: I agree with it, but kind of, we're using session widely in this way, for all packages talk to govmomi.
This is a reasonable claim drive us to refactor session I think, but I would prefer to do it separately.

File: pkg/vsphere/tasks/waiter.go (hunk 6 above)
Comment (emlin) [inline pkg/vsphere/tasks/waiter.go]: Initially I'm doing that way, but I think invalid state will be a much more rare condition than any others. In fact, we'll never have a nil target, unless in test case.

File: pkg/vsphere/vm/vm.go (hunk 10 above)
Comment (emlin) [inline pkg/vsphere/vm/vm.go]: vm.Client() is not a valid method. but I can call vm.Vim25()

File: pkg/vsphere/vm/vm.go (hunk 13 above)
Comment (emlin) [inline pkg/vsphere/vm/vm.go]: Here, I'm trying not to impact the virtual machine object itself, and reset the vm reference.
From caller point of view, it might cache the object, and continue to use that after this waitforresult method.
But inside of waitforresult method, we actually changed the VM itself from VC, so the idea to change object as less as possible. Ideally I want to set object reference only back to the object, but no such method.
NewVirtualMachine is to create a common object only, but it might cache other values after some operation, we don't want to loss those caches just inside of one method call blindly.

File: pkg/vsphere/tasks/waiter.go (hunk 9 above)
Comment (hmahmood) [inline pkg/vsphere/tasks/waiter.go]: You are calling `needsFix` again after this; I thought we could get rid of the extra call this way.

File: pkg/vsphere/tasks/waiter.go (hunk 7 above)
Comment (hmahmood) [inline pkg/vsphere/tasks/waiter.go]: I think its better if we do it now, since changes to WaitForResult result in a lot of churn.

File: lib/portlayer/exec/exec.go [hunk 15]
```diff
@@ -152,8 +158,36 @@ func eventCallback(ie events.Event) {
...
+	case events.ContainerRegistered:
+		moref := new(types.ManagedObjectReference)
+		if ok := moref.FromString(ie.Reference()); !ok {
+			log.Errorf("Failed to get event VM mobref: %s", ie.Reference())
+			return
+		}
+		var vm mo.VirtualMachine
+
+		// current attributes we care about
+		attrib := []string{"config", "runtime.powerState", "summary", "resourcePool"}
```
Comment (hmahmood) [inline lib/portlayer/exec/exec.go]: Why are we retrieving config, runtime.powerState and summary? Seems like only resourcePool is used below.

File: lib/portlayer/exec/exec.go [hunk 16]
```diff
@@ -152,8 +158,36 @@ func eventCallback(ie events.Event) {
...
 
+		// populate the vm properties
+		ctx := context.Background()
+		if err := eventSession.RetrieveOne(ctx, *moref, attrib, &vm); err != nil {
+			log.Errorf("Failed to query registered vm object %s: %s", ie.Reference(), err)
+			return
+		}
+		if *vm.ResourcePool != Config.ResourcePool.Reference() {
```
Comment (hmahmood) [inline lib/portlayer/exec/exec.go]: This seems to be very specific to VM objects. Maybe we should move this "fix up" to the VirtualMachine class (via a WaitForResult or similar method), and not do it here? Adding these two parameters really makes this function very confusing, and ends up with WaitForResult doing to much imho.

File: pkg/vsphere/vm/vm.go (hunk 13 above)
Comment (hmahmood) [inline pkg/vsphere/vm/vm.go]: How about we just return a new VirtualMachine object? The old one would be defunct after this call.

Comment (hmahmood) [inline pkg/vsphere/vm/vm.go]: What do you think about just returning a new VirtualMachine object? Coupled with a WaitForResult method/func in this file, that could work. The existing VirtualMachine object would be defunct after this call.

File: pkg/vsphere/tasks/waiter.go (hunk 8 above)
Comment (emlin) [inline pkg/vsphere/tasks/waiter.go]: The change itself turns to be much more confusing than everybody thought. But the result is that we have to do it in tasks.WaitForResult, cause every VM operation might end up with this failure, and the solution is not only to fix the error, but also retry task.
If we move the fix out of WaitForResult, we'll need to modify every VM operation call, patch with this workaround, that's not what we should do.

File: pkg/vsphere/tasks/waiter.go (hunk 9 above)
Comment (emlin) [inline pkg/vsphere/tasks/waiter.go]: Either way works. If you prefer, I can change it.

File: pkg/vsphere/tasks/waiter.go (hunk 8 above)
Comment (hmahmood) [inline pkg/vsphere/tasks/waiter.go]: I meant to say add it to the VirtualMachine class or a helper function in vm.go, that way we use that for VM objects only.

File: pkg/vsphere/tasks/waiter.go (hunk 7 above)
Comment (emlin) [inline pkg/vsphere/tasks/waiter.go]: We still rely on session to get VMFolder, if it's ESXi, or we do not use resource pool. Remove Session cannot simply make it easy.

File: lib/portlayer/exec/exec.go (hunk 15 above)
Comment (emlin) [inline lib/portlayer/exec/exec.go]: I knew that. But I need one more attribute for the resource pool. If call populateVMAttributes, I'll need to call property collector twice.

File: lib/portlayer/exec/exec.go (hunk 16 above)
Comment (emlin) [inline lib/portlayer/exec/exec.go]: which kind of helper? I assume that will include the VM property query, right?

**Comment by andrewtchin** [comment]
lgtm

File: lib/portlayer/exec/exec.go [hunk 17]
```diff
@@ -24,18 +24,24 @@ import (
...
 )
 
-var initializer sync.Once
+var (
+	initializer  sync.Once
+	eventSession *session.Session
+)
 
 func Init(ctx context.Context, sess *session.Session, source extraconfig.DataSource, _ extraconfig.DataSink) error {
 	var err error
```
Comment (hmahmood) [inline lib/portlayer/exec/exec.go]: Unrelated... but the error we are returning is going to be always nil after the first called to Init().

File: pkg/vsphere/vm/vm.go [hunk 18]
```diff
@@ -427,3 +425,110 @@ func (vm *VirtualMachine) UpgradeInProgress(ctx context.Context, upgradePrefix s
...
+		log.Errorf("Unable to register VM %q back: %s", name, err)
+		return err
+	}
+	info, err := task.WaitForResult(ctx, nil)
+	if err != nil {
+		return err
+	}
+	// re-register vm will change vm reference, so reset the object reference here
+	if info.Error != nil {
+		return errors.New(info.Error.LocalizedMessage)
```
Comment (hmahmood) [inline pkg/vsphere/vm/vm.go]: We probably should just return the `error` as is from the `WaitForResult` call.

**Comment by hmahmood** [comment]
Minor comments ... lgtm

File: pkg/vsphere/vm/vm.go (hunk 18 above)
Comment (emlin) [inline pkg/vsphere/vm/vm.go]: In this case, err is nil, but info.Error is not. I doubt this will happen, but if it does, this error might be better.

File: lib/portlayer/exec/exec.go [hunk 19]
```diff
@@ -111,8 +114,10 @@ func Init(ctx context.Context, sess *session.Session, source extraconfig.DataSou
 			return
 		}
 	})
-
-	return err
+	if err != nil {
```
Comment (hmahmood) [inline lib/portlayer/exec/exec.go]: This needs to be inside the `func()` above.

Comment (emlin) [inline lib/portlayer/exec/exec.go]: okey

File: lib/portlayer/exec/exec.go [hunk 20]
```diff
@@ -187,15 +202,15 @@ func eventCallback(ie events.Event) {
 	return
 }
 
-func isManagedbyVCH(moref types.ManagedObjectReference) bool {
+func isManagedbyVCH(moref types.ManagedObjectReference, sess *session.Session) bool {
```
Comment (emlin) [inline lib/portlayer/exec/exec.go]: fine

**Comment by emlin** [comment]
♻️
